### DISCUSSION TITLE
How to I connect Django app to OCI Autonomous DB?

### DISCUSSION THREAD

**Discussion by mikekenn** [oracle/python-oracledb#155] [Q&A] [answered]
Im following along in the [Django documentation](https://docs.djangoproject.com/en/4.1/intro/tutorial01/) and I want to connect the application for a Free Tier OCI DB that I have running.

I am able to connect to the DB via VS Code and perform all the expected actions. But I would like it, so that when I do `python manage.py migrate` all the tables get created in the Oracle DB rather than the provided sqlite db. 

I would also like to be able to run packages/procedures from the app as well. Do I still have to create a connection to the database every time I want to do a query/insert/update? Something like this?

```
connection = oracledb.connect(
                              user="USERNAME",
                              password="PASSWORD",
                              dsn="DSN_VALUE_FROM_OCI_CONNECTION_UI",
                              config_dir="/Location/ToApp/Wallet",
                              wallet_location="/Location/ToApp/Wallet",
                              wallet_password=pw)

with connection.cursor() as cursor:

    cursor.execute("""
        begin
            execute immediate 'drop table todoitem';
            exception when others then if sqlcode <> -942 then raise; end if;
        end;""")

    cursor.execute("""
        create table todoitem (
            id number generated always as identity,
            description varchar2(4000),
            creation_ts timestamp with time zone default current_timestamp,
            done number(1,0),
            primary key (id))""")

    print("Table created")
```

I have changed the section in settings.py to the following & I have OracleDB installed

```
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': 'NAME_FOUND_IN_ORACLE_CLOUD_UI',
        'USER': 'USERNAME',
        'PASSWORD': 'USER_PASSWORD',
        'HOST': 'Value_In_TNSNAMES.ORA_FILE',
        'PORT': 'Value_In_TNSNAMES.ORA_FILE', 
    }
}
```

Is there something I am missing? Also, I am working on a "3 GHz 6-Core Intel Core i5" desktop & a "Apple M1" laptop. Guidance on how to get this to work on both machines would be greatly appreciated!

Please let me know if there is more information that is required!

**Comment by cjbj** [comment]
@suraj-ora-2020 can you comment?

**Comment by suraj-ora-2020** [comment, marked as answer]
Since you are able to connect you can use the following command to migrate all the tables.
python manage.py inspectdb > myapp/models.py
This will create the Objects based on the relational db.

**Reply by mikekenn** [reply]
Thank you! @suraj-ora-2020

**Reply by mikekenn** [reply]
@suraj-ora-2020 || @cjbj Is there documentation on how to connect on Apple Silicon M1 without using Rosetta2

**Reply by mikekenn** [reply]
Also, @suraj-ora-2020 || @cjbj 

When I am connecting on an MBP - Apple M1. Do I still need the following in settings.py?

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': 'NAME_FOUND_IN_ORACLE_CLOUD_UI',
        'USER': 'USERNAME',
        'PASSWORD': 'USER_PASSWORD',
        'HOST': 'Value_In_TNSNAMES.ORA_FILE',
        'PORT': 'Value_In_TNSNAMES.ORA_FILE', 
    }
}

**Reply by suraj-ora-2020** [reply]
Yes you need this configuration to be able to connect to database using Django.

**Comment by cjbj** [comment]
Future readers may be interested in:

- https://cjones-oracle.medium.com/django-5-and-oracle-autonomous-database-fc73207bf984
- https://cjones-oracle.medium.com/django-5-0-supports-python-oracledb-natively-46856cb68658
- https://cjones-oracle.medium.com/using-python-oracledb-1-0-with-sqlalchemy-pandas-django-and-flask-5d84e910cb19

And since this discussion was first initiated, Instant Client for macOS ARM64 has been released if you need to use python-oracledb Thick mode: https://medium.com/@cjones-oracle/installing-oracle-instant-client-for-apple-macos-arm64-m1-m2-m3-2c81f246feb9
